File: packages/shared-types/src/schemas.ts
export type JsonSchema = Record<string, unknown>;

export const markerStatusSchema: JsonSchema = {
  $id: "MarkerStatus",
  type: "string",
  enum: ["human_draft", "ai_detected", "ai_review", "human_confirmed", "human_corrected", "rejected"],
};

export const markerPointTypeSchema: JsonSchema = {
  $id: "MarkerPointType",
  type: "string",
  enum: ["center", "top_left"],
};

export const candidateKindSchema: JsonSchema = {
  $id: "CandidateKind",
  type: "string",
  enum: ["circle", "box", "text"],
};

export const candidateReviewStatusSchema: JsonSchema = {
  $id: "CandidateReviewStatus",
  type: "string",
  enum: ["pending", "accepted", "rejected"],
};

export const pipelineConflictTypeSchema: JsonSchema = {
  $id: "PipelineConflictType",
  type: "string",
  enum: ["candidate_ambiguity", "association_ambiguity", "duplicate_label_nearby", "missing_vocab_label"],
};

export const pipelineConflictSeveritySchema: JsonSchema = {
  $id: "PipelineConflictSeverity",
  type: "string",
  enum: ["warning", "error"],
};

export const sessionCommandTypeSchema: JsonSchema = {
  $id: "SessionCommandType",
  type: "string",
  enum: [
    "set_viewport",
    "pan_viewport",
    "zoom_to_region",
    "place_marker",
    "move_marker",
    "update_marker",
    "confirm_marker",
    "reject_marker",
    "delete_marker",
    "clear_markers",
    "reject_candidate",
  ],
};

export const actorSchema: JsonSchema = {
  $id: "Actor",
  type: "string",
  enum: ["human", "ai", "system"],
};

export const viewportSchema: JsonSchema = {
  $id: "Viewport",
  type: "object",
  required: ["centerX", "centerY", "zoom"],
  properties: {
    centerX: { type: "number" },
    centerY: { type: "number" },
    zoom: { type: "number", minimum: 0.1 },
  },
  additionalProperties: false,
};

export const documentAssetSchema: JsonSchema = {
  $id: "DocumentAsset",
  type: "object",
  required: ["documentId", "fileName", "contentType", "sizeBytes", "width", "height", "storageUrl", "uploadedAt"],
  properties: {
    documentId: { type: "string" },
    fileName: { type: "string" },
    contentType: { type: "string" },
    sizeBytes: { type: "integer", minimum: 1 },
    width: { type: "integer", minimum: 1 },
    height: { type: "integer", minimum: 1 },
    storageUrl: { type: "string" },
    uploadedAt: { type: "string", format: "date-time" },
  },
  additionalProperties: false,
};

export const markerSchema: JsonSchema = {
  $id: "Marker",
  type: "object",
  required: [
    "markerId",
    "label",
    "x",
    "y",
    "pointType",
    "status",
    "confidence",
    "createdBy",
    "updatedBy",
    "createdAt",
    "updatedAt",
  ],
  properties: {
    markerId: { type: "string" },
    label: { type: ["string", "null"] },
    x: { type: "number", minimum: 0 },
    y: { type: "number", minimum: 0 },
    pointType: { $ref: "MarkerPointType" },
    status: { $ref: "MarkerStatus" },
    confidence: { type: ["number", "null"], minimum: 0, maximum: 1 },
    createdBy: { $ref: "Actor" },
    updatedBy: { $ref: "Actor" },
    createdAt: { type: "string", format: "date-time" },
    updatedAt: { type: "string", format: "date-time" },
  },
  additionalProperties: false,
};

export const calloutCandidateSchema: JsonSchema = {
  $id: "CalloutCandidate",
  type: "object",
  required: [
    "candidateId",
    "kind",
    "centerX",
    "centerY",
    "bboxX",
    "bboxY",
    "bboxWidth",
    "bboxHeight",
    "score",
    "cropUrl",
    "suggestedLabel",
    "suggestedConfidence",
    "suggestedSource",
    "topologyScore",
    "topologySource",
    "leaderAnchorX",
    "leaderAnchorY",
    "reviewStatus",
    "conflictGroup",
    "conflictCount",
    "createdAt",
    "updatedAt",
  ],
  properties: {
    candidateId: { type: "string" },
    kind: { $ref: "CandidateKind" },
    centerX: { type: "number", minimum: 0 },
    centerY: { type: "number", minimum: 0 },
    bboxX: { type: "number", minimum: 0 },
    bboxY: { type: "number", minimum: 0 },
    bboxWidth: { type: "number", exclusiveMinimum: 0 },
    bboxHeight: { type: "number", exclusiveMinimum: 0 },
    score: { type: "number", minimum: 0 },
    cropUrl: { type: ["string", "null"] },
    suggestedLabel: { type: ["string", "null"] },
    suggestedConfidence: { type: ["number", "null"], minimum: 0, maximum: 1 },
    suggestedSource: { type: ["string", "null"] },
    topologyScore: { type: ["number", "null"], minimum: 0, maximum: 1 },
    topologySource: { type: ["string", "null"] },
    leaderAnchorX: { type: ["number", "null"], minimum: 0 },
    leaderAnchorY: { type: ["number", "null"], minimum: 0 },
    reviewStatus: { $ref: "CandidateReviewStatus" },
    conflictGroup: { type: ["string", "null"] },
    conflictCount: { type: "integer", minimum: 0 },
    createdAt: { type: "string", format: "date-time" },
    updatedAt: { type: "string", format: "date-time" },
  },
  additionalProperties: false,
};

export const candidateAssociationSchema: JsonSchema = {
  $id: "CandidateAssociation",
  type: "object",
  required: [
    "associationId",
    "shapeCandidateId",
    "textCandidateId",
    "shapeKind",
    "label",
    "score",
    "geometryScore",
    "topologyScore",
    "source",
    "leaderAnchorX",
    "leaderAnchorY",
    "bboxX",
    "bboxY",
    "bboxWidth",
    "bboxHeight",
  ],
  properties: {
    associationId: { type: "string" },
    shapeCandidateId: { type: "string" },
    textCandidateId: { type: "string" },
    shapeKind: { $ref: "CandidateKind" },
    label: { type: "string" },
    score: { type: "number", minimum: 0, maximum: 1 },
    geometryScore: { type: "number", minimum: 0, maximum: 1 },
    topologyScore: { type: ["number", "null"], minimum: 0, maximum: 1 },
    source: { type: "string" },
    leaderAnchorX: { type: ["number", "null"], minimum: 0 },
    leaderAnchorY: { type: ["number", "null"], minimum: 0 },
    bboxX: { type: ["number", "null"], minimum: 0 },
    bboxY: { type: ["number", "null"], minimum: 0 },
    bboxWidth: { type: ["number", "null"], minimum: 0 },
    bboxHeight: { type: ["number", "null"], minimum: 0 },
  },
  additionalProperties: false,
};

export const pageVocabularyEntrySchema: JsonSchema = {
  $id: "PageVocabularyEntry",
  type: "object",
  required: [
    "label",
    "normalizedLabel",
    "occurrences",
    "maxConfidence",
    "sources",
    "bboxX",
    "bboxY",
    "bboxWidth",
    "bboxHeight",
  ],
  properties: {
    label: { type: "string" },
    normalizedLabel: { type: "string" },
    occurrences: { type: "integer", minimum: 0 },
    maxConfidence: { type: ["number", "null"], minimum: 0, maximum: 1 },
    sources: { type: "array", items: { type: "string" } },
    bboxX: { type: ["number", "null"], minimum: 0 },
    bboxY: { type: ["number", "null"], minimum: 0 },
    bboxWidth: { type: ["number", "null"], minimum: 0 },
    bboxHeight: { type: ["number", "null"], minimum: 0 },
  },
  additionalProperties: false,
};

export const pipelineConflictSchema: JsonSchema = {
  $id: "PipelineConflict",
  type: "object",
  required: [
    "conflictId",
    "type",
    "severity",
    "label",
    "message",
    "candidateIds",
    "markerIds",
    "relatedLabels",
    "bboxX",
    "bboxY",
    "bboxWidth",
    "bboxHeight",
  ],
  properties: {
    conflictId: { type: "string" },
    type: { $ref: "PipelineConflictType" },
    severity: { $ref: "PipelineConflictSeverity" },
    label: { type: ["string", "null"] },
    message: { type: "string" },
    candidateIds: { type: "array", items: { type: "string" } },
    markerIds: { type: "array", items: { type: "string" } },
    relatedLabels: { type: "array", items: { type: "string" } },
    bboxX: { type: ["number", "null"], minimum: 0 },
    bboxY: { type: ["number", "null"], minimum: 0 },
    bboxWidth: { type: ["number", "null"], minimum: 0 },
    bboxHeight: { type: ["number", "null"], minimum: 0 },
  },
  additionalProperties: false,
};

export const actionLogEntrySchema: JsonSchema = {
  $id: "ActionLogEntry",
  type: "object",
  required: ["actionId", "actor", "type", "createdAt", "payload"],
  properties: {
    actionId: { type: "string" },
    actor: { $ref: "Actor" },
    type: { type: "string" },
    createdAt: { type: "string", format: "date-time" },
    payload: { type: "object" },
  },
  additionalProperties: false,
};

export const sessionSummarySchema: JsonSchema = {
  $id: "SessionSummary",
  type: "object",
  required: ["totalMarkers", "aiDetected", "aiReview", "humanConfirmed", "humanCorrected", "rejected"],
  properties: {
    totalMarkers: { type: "integer", minimum: 0 },
    aiDetected: { type: "integer", minimum: 0 },
    aiReview: { type: "integer", minimum: 0 },
    humanConfirmed: { type: "integer", minimum: 0 },
    humanCorrected: { type: "integer", minimum: 0 },
    rejected: { type: "integer", minimum: 0 },
  },
  additionalProperties: false,
};

export const annotationSessionSchema: JsonSchema = {
  $id: "AnnotationSession",
  type: "object",
  required: [
    "sessionId",
    "title",
    "state",
    "document",
    "viewport",
    "candidates",
    "candidateAssociations",
    "pageVocabulary",
    "missingLabels",
    "pipelineConflicts",
    "markers",
    "actionLog",
    "summary",
    "createdAt",
    "updatedAt",
  ],
  properties: {
    sessionId: { type: "string" },
    title: { type: "string" },
    state: { type: "string", enum: ["draft", "ready"] },
    document: { anyOf: [{ $ref: "DocumentAsset" }, { type: "null" }] },
    viewport: { $ref: "Viewport" },
    candidates: { type: "array", items: { $ref: "CalloutCandidate" } },
    candidateAssociations: { type: "array", items: { $ref: "CandidateAssociation" } },
    pageVocabulary: { type: "array", items: { $ref: "PageVocabularyEntry" } },
    missingLabels: { type: "array", items: { type: "string" } },
    pipelineConflicts: { type: "array", items: { $ref: "PipelineConflict" } },
    markers: { type: "array", items: { $ref: "Marker" } },
    actionLog: { type: "array", items: { $ref: "ActionLogEntry" } },
    summary: { $ref: "SessionSummary" },
    createdAt: { type: "string", format: "date-time" },
    updatedAt: { type: "string", format: "date-time" },
  },
  additionalProperties: false,
};

export const apiContracts = {
  CreateSessionResponse: {
    $id: "CreateSessionResponse",
    type: "object",
    required: ["session"],
    properties: {
      session: { $ref: "AnnotationSession" },
    },
    additionalProperties: false,
  },
  SessionListResponse: {
    $id: "SessionListResponse",
    type: "object",
    required: ["sessions"],
    properties: {
      sessions: {
        type: "array",
        items: {
          type: "object",
          required: ["sessionId", "title", "state", "documentName", "markerCount", "updatedAt"],
          properties: {
            sessionId: { type: "string" },
            title: { type: "string" },
            state: { type: "string", enum: ["draft", "ready"] },
            documentName: { type: ["string", "null"] },
            markerCount: { type: "integer", minimum: 0 },
            updatedAt: { type: "string", format: "date-time" },
          },
          additionalProperties: false,
        },
      },
    },
    additionalProperties: false,
  },
  UploadDocumentResponse: {
    $id: "UploadDocumentResponse",
    type: "object",
    required: ["session"],
    properties: {
      session: { $ref: "AnnotationSession" },
    },
    additionalProperties: false,
  },
  SessionCommandResponse: {
    $id: "SessionCommandResponse",
    type: "object",
    required: ["session"],
    properties: {
      session: { $ref: "AnnotationSession" },
    },
    additionalProperties: false,
  },
};

File: services/inference/app/models/schemas.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field


def to_camel(value: str) -> str:
    head, *tail = value.split("_")
    return head + "".join(part.capitalize() for part in tail)


class ApiModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)


class SessionState(str, Enum):
    DRAFT = "draft"
    READY = "ready"


class Actor(str, Enum):
    HUMAN = "human"
    AI = "ai"
    SYSTEM = "system"


class MarkerStatus(str, Enum):
    HUMAN_DRAFT = "human_draft"
    AI_DETECTED = "ai_detected"
    AI_REVIEW = "ai_review"
    HUMAN_CONFIRMED = "human_confirmed"
    HUMAN_CORRECTED = "human_corrected"
    REJECTED = "rejected"


class MarkerPointType(str, Enum):
    CENTER = "center"
    TOP_LEFT = "top_left"


class CandidateKind(str, Enum):
    CIRCLE = "circle"
    BOX = "box"
    TEXT = "text"


class CandidateReviewStatus(str, Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    REJECTED = "rejected"


class PipelineConflictType(str, Enum):
    CANDIDATE_AMBIGUITY = "candidate_ambiguity"
    ASSOCIATION_AMBIGUITY = "association_ambiguity"
    DUPLICATE_LABEL_NEARBY = "duplicate_label_nearby"
    MISSING_VOCAB_LABEL = "missing_vocab_label"


class PipelineConflictSeverity(str, Enum):
    WARNING = "warning"
    ERROR = "error"


class ActionType(str, Enum):
    SESSION_CREATED = "session_created"
    DOCUMENT_UPLOADED = "document_uploaded"
    CANDIDATES_DETECTED = "candidates_detected"
    AUTO_ANNOTATION_COMPLETED = "auto_annotation_completed"
    VIEWPORT_SET = "viewport_set"
    VIEWPORT_PANNED = "viewport_panned"
    VIEWPORT_ZOOMED = "viewport_zoomed"
    CANDIDATE_REJECTED = "candidate_rejected"
    MARKER_CREATED = "marker_created"
    MARKER_MOVED = "marker_moved"
    MARKER_UPDATED = "marker_updated"
    MARKER_CONFIRMED = "marker_confirmed"
    MARKER_REJECTED = "marker_rejected"
    MARKER_DELETED = "marker_deleted"
    MARKERS_CLEARED = "markers_cleared"


class SessionCommandType(str, Enum):
    SET_VIEWPORT = "set_viewport"
    PAN_VIEWPORT = "pan_viewport"
    ZOOM_TO_REGION = "zoom_to_region"
    PLACE_MARKER = "place_marker"
    MOVE_MARKER = "move_marker"
    UPDATE_MARKER = "update_marker"
    CONFIRM_MARKER = "confirm_marker"
    REJECT_MARKER = "reject_marker"
    DELETE_MARKER = "delete_marker"
    CLEAR_MARKERS = "clear_markers"
    REJECT_CANDIDATE = "reject_candidate"


class Viewport(ApiModel):
    center_x: float = 0
    center_y: float = 0
    zoom: float = Field(default=1, gt=0)


class DocumentAsset(ApiModel):
    document_id: str = Field(default_factory=lambda: str(uuid4()))
    file_name: str
    content_type: str
    size_bytes: int = Field(ge=1)
    width: int = Field(ge=1)
    height: int = Field(ge=1)
    storage_url: str
    uploaded_at: datetime = Field(default_factory=datetime.utcnow)


class Marker(ApiModel):
    marker_id: str = Field(default_factory=lambda: str(uuid4()))
    label: Optional[str] = None
    x: float = Field(ge=0)
    y: float = Field(ge=0)
    point_type: MarkerPointType = MarkerPointType.CENTER
    status: MarkerStatus
    confidence: Optional[float] = Field(default=None, ge=0, le=1)
    created_by: Actor
    updated_by: Actor
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class CalloutCandidate(ApiModel):
    candidate_id: str = Field(default_factory=lambda: str(uuid4()))
    kind: CandidateKind
    center_x: float = Field(ge=0)
    center_y: float = Field(ge=0)
    bbox_x: float = Field(ge=0)
    bbox_y: float = Field(ge=0)
    bbox_width: float = Field(gt=0)
    bbox_height: float = Field(gt=0)
    score: float = Field(ge=0)
    crop_url: Optional[str] = None
    suggested_label: Optional[str] = None
    suggested_confidence: Optional[float] = Field(default=None, ge=0, le=1)
    suggested_source: Optional[str] = None
    topology_score: Optional[float] = Field(default=None, ge=0, le=1)
    topology_source: Optional[str] = None
    leader_anchor_x: Optional[float] = Field(default=None, ge=0)
    leader_anchor_y: Optional[float] = Field(default=None, ge=0)
    review_status: CandidateReviewStatus = CandidateReviewStatus.PENDING
    conflict_group: Optional[str] = None
    conflict_count: int = Field(default=0, ge=0)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class CandidateAssociation(ApiModel):
    association_id: str = Field(default_factory=lambda: str(uuid4()))
    shape_candidate_id: str
    text_candidate_id: str
    shape_kind: CandidateKind
    label: str
    score: float = Field(ge=0, le=1)
    geometry_score: float = Field(ge=0, le=1)
    topology_score: Optional[float] = Field(default=None, ge=0, le=1)
    source: str
    leader_anchor_x: Optional[float] = Field(default=None, ge=0)
    leader_anchor_y: Optional[float] = Field(default=None, ge=0)
    bbox_x: Optional[float] = Field(default=None, ge=0)
    bbox_y: Optional[float] = Field(default=None, ge=0)
    bbox_width: Optional[float] = Field(default=None, ge=0)
    bbox_height: Optional[float] = Field(default=None, ge=0)


class PageVocabularyEntry(ApiModel):
    label: str
    normalized_label: str
    occurrences: int = Field(default=0, ge=0)
    max_confidence: Optional[float] = Field(default=None, ge=0, le=1)
    sources: list[str] = Field(default_factory=list)
    bbox_x: Optional[float] = Field(default=None, ge=0)
    bbox_y: Optional[float] = Field(default=None, ge=0)
    bbox_width: Optional[float] = Field(default=None, ge=0)
    bbox_height: Optional[float] = Field(default=None, ge=0)


class PipelineConflict(ApiModel):
    conflict_id: str = Field(default_factory=lambda: str(uuid4()))
    type: PipelineConflictType
    severity: PipelineConflictSeverity
    label: Optional[str] = None
    message: str
    candidate_ids: list[str] = Field(default_factory=list)
    marker_ids: list[str] = Field(default_factory=list)
    related_labels: list[str] = Field(default_factory=list)
    bbox_x: Optional[float] = Field(default=None, ge=0)
    bbox_y: Optional[float] = Field(default=None, ge=0)
    bbox_width: Optional[float] = Field(default=None, ge=0)
    bbox_height: Optional[float] = Field(default=None, ge=0)


class ActionLogEntry(ApiModel):
    action_id: str = Field(default_factory=lambda: str(uuid4()))
    actor: Actor
    type: ActionType
    created_at: datetime = Field(default_factory=datetime.utcnow)
    payload: Dict[str, Any] = Field(default_factory=dict)


class SessionSummary(ApiModel):
    total_markers: int = 0
    ai_detected: int = 0
    ai_review: int = 0
    human_confirmed: int = 0
    human_corrected: int = 0
    rejected: int = 0


class AnnotationSession(ApiModel):
    session_id: str = Field(default_factory=lambda: str(uuid4()))
    title: str = "Untitled session"
    state: SessionState = SessionState.DRAFT
    document: Optional[DocumentAsset] = None
    viewport: Viewport = Field(default_factory=Viewport)
    candidates: list[CalloutCandidate] = Field(default_factory=list)
    candidate_associations: list[CandidateAssociation] = Field(default_factory=list)
    page_vocabulary: list[PageVocabularyEntry] = Field(default_factory=list)
    missing_labels: list[str] = Field(default_factory=list)
    pipeline_conflicts: list[PipelineConflict] = Field(default_factory=list)
    markers: list[Marker] = Field(default_factory=list)
    action_log: list[ActionLogEntry] = Field(default_factory=list)
    summary: SessionSummary = Field(default_factory=SessionSummary)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class SessionListItem(ApiModel):
    session_id: str
    title: str
    state: SessionState
    document_name: Optional[str] = None
    marker_count: int = 0
    updated_at: datetime


class CreateSessionRequest(ApiModel):
    title: Optional[str] = None


class CreateSessionResponse(ApiModel):
    session: AnnotationSession


class SessionListResponse(ApiModel):
    sessions: list[SessionListItem]


class UploadDocumentResponse(ApiModel):
    session: AnnotationSession


class SessionCommandRequest(ApiModel):
    type: SessionCommandType
    actor: Actor = Actor.HUMAN
    marker_id: Optional[str] = None
    candidate_id: Optional[str] = None
    label: Optional[str] = None
    point_type: Optional[MarkerPointType] = None
    status: Optional[MarkerStatus] = None
    confidence: Optional[float] = Field(default=None, ge=0, le=1)
    x: Optional[float] = None
    y: Optional[float] = None
    width: Optional[float] = None
    height: Optional[float] = None
    center_x: Optional[float] = None
    center_y: Optional[float] = None
    zoom: Optional[float] = Field(default=None, gt=0)
    delta_x: Optional[float] = None
    delta_y: Optional[float] = None


class SessionCommandResponse(ApiModel):
    session: AnnotationSession
